File: Prob331/Solution.cpp
// Approach 1:-

class Solution
{
private:
    int index = 0;
    bool res = true;

public:
    bool isValidSerialization(string preorder)
    {

        preOrder(preorder);

        return index >= preorder.size() && res ? true : false;
    }

    void preOrder(string &preorder)
    {
        if (index >= preorder.size())
        {
            res = false;
            return;
        }
        int i = index;

        while (i < preorder.size() && preorder[i] != ',')
            i++;

        string temp = preorder.substr(index, i - index);
        index = i + 1;

        if (temp == "#")
            return;

        preOrder(preorder);

        preOrder(preorder);
    }
};

File: Prob331/Solution.java
// Approach 1:-

class Solution {
    private int index=0;
    private boolean res=true;
    
    public boolean isValidSerialization(String preorder) {
        preOrder(preorder);
        
        return index>=preorder.length() && res;
    }
    
    void preOrder(String preorder) {
        if(index>=preorder.length()) {
            res=false;
            return ;
        }
        
        int i=index;
        
        while(i<preorder.length() && preorder.charAt(i)!=',')
            i++;
        
        String temp=preorder.substring(index,i);
        index=i+1;
        
        if(temp.equals("#"))
            return ;
        
        preOrder(preorder);
        preOrder(preorder);
    }
}